salt/master.py

```diff
@@ -1979,6 +1980,40 @@ def __init__(self, opts, key):
         self.masterapi = salt.daemons.masterapi.LocalFuncs(opts, key)
         self.channels = []
 
+    def _list_valid_minions(self, clear_load):
+        delimiter = clear_load.get("kwargs", {}).get("delimiter", DEFAULT_TARGET_DELIM)
+        _res = self.ckminions.check_minions(
+            clear_load["tgt"], clear_load.get("tgt_type", "glob"), delimiter
+        )
+        tgt_minions = _res["minions"]
+        valid_minions = set()
+        for v in clear_load["valid_tgt"]:
+            _res = self.ckminions.check_minions(
+                v, clear_load.get("valid_tgt_type", "glob"), delimiter
+            )
+            valid_minions.update(_res["minions"])
+        ret = {"valid_minions": list(valid_minions), "tgt_minions": tgt_minions}
+        event_ret = {
+            "fun": "sys.tgt_match_ret",
+            "tgt": clear_load["tgt"],
+            "tgt_type": "glob",
+            "arg": ret,
+            "user": "root",
+            "ret": "",
+            "kwargs": {"blerp": self.opts["id"]},
+            "jid": clear_load.get("jid"),
+        }
+
+        missing = []
+        jid = self._prep_jid(event_ret, {})
+        payload = self._prep_pub([], jid, event_ret, {}, missing)
+        self.event.fire_event(payload, tagify([jid, "tgt_return"]))
+        # self.event.fire_event(
```

Comment:
See my other comment - one or the other :joy: 

I haven't gone back and tested, but now that my tests are all passing, and passing in ~5m it's going to be a *lot* easier to test, but do you know if it matters if I just fire an event vs create the payload with prepping the jid, etc?

My gut says that it would actually be better to just fire an event rather than do all the job bits... but again I haven't tested it yet. I have a vague memory that at one point I tried it and I could see the event over on the `_collect_syndics` event watcher.

Reply:
Okay, so experiments verify that I can, in fact, avoid doing the prep_pub and prep_jid steps and simply fire off an event! Seems to simplify things quite a bit, too.